race-buffer/src/writer.rs

```diff
@@ -0,0 +1,252 @@
+            Self::Entry(e) => e,
+        }
+    }
+}
+
+/// An entry or double entry that has just been overwritten
+pub enum OverwrittenEntry<E>
+where
+    E: Entry,
+{
+    None,
+    Single(E),
+    Double(E, E),
+}
+
+#[derive(Debug)]
+#[repr(C)]
+/// Struct used to write to buffer
+pub struct RaceBuffer<'a, E>
+where
+    E: Entry,
+{
+    /// Write cursor: Cursor pointing to where the next entry will be written
+    wcurs: usize,
+    /// Overwrite cursor: Entries in front of wcurs but behind owcurs are in the process of being overwritten
+    owcurs: usize,
+    /// Backing storage
+    storage: &'a mut [MaybeUninit<E>],
+    /// Indicates if backing storage should be truncated to a power of 2 length
+    /// in order to use optimized indexing
+    use_base_2_indexing: bool,
+}
+
+impl<'a, E> RaceBuffer<'a, E>
+where
+    E: Entry,
+{
+    /// Create new RaceBuffer. Returns error if storage size is not power of 2
+    pub fn new(storage: &'a mut [MaybeUninit<E>]) -> Result<RaceBuffer<'a, E>, SizeError> {
+        let use_base_2_indexing = storage.len().is_power_of_two();
+        if storage.len() < MIN_STORAGE_CAP {
+            Err(SizeError)
+        } else {
+            Ok(RaceBuffer {
+                wcurs: 0,
+                owcurs: storage.len(),
+                storage,
+                use_base_2_indexing,
+            })
+        }
+    }
+
+    /// Create new RaceBuffer with properly aligned backing storage
+    #[inline]
+    pub fn new_from_bytes(bytes: &'a mut [u8]) -> Result<RaceBuffer<'a, E>, SizeError> {
+        let (_prefix, buf, _suffix) = Self::align_from_bytes(bytes);
+        buf
+    }
+
+    /// Create new RaceBuffer with properly aligned backing storage, return unused bytes
+    #[inline]
+    pub fn align_from_bytes(
+        bytes: &'a mut [u8],
+    ) -> (
+        &'a mut [u8],
+        Result<RaceBuffer<'a, E>, SizeError>,
+        &'a mut [u8],
+    ) {
+        // Safe because storage is treated as uninit after transmutation
+        let (prefix, storage, suffix) = unsafe { bytes.align_to_mut() };
+        (prefix, RaceBuffer::new(storage), suffix)
+    }
+
+    /// Write single entry to buffer, returning overwritten entry
+    pub fn write(&mut self, entry: E) -> OverwrittenEntry<E> {
+        let mut overwritten = OverwrittenEntry::None;
+        if self.wcurs == self.owcurs {
+            // Reading storage is safe after checking if buffer has looped at least once
+            if self.wcurs >= self.storage.len()
+                && unsafe { self.read_storage(self.wcurs).is_prefix() }
+            {
+                // If overwriting double, increment overwrite cursor by 2
+                overwritten = unsafe {
+                    OverwrittenEntry::Double(
+                        self.read_storage(self.owcurs),
+                        self.read_storage(self.owcurs + 1),
+                    )
+                };
+                self.owcurs += 2;
```

Comment:
Am I correct in assuming the intent is that `owcurs` and `wcurs` are supposed to wrap when they hit `usize::MAX`?  

If so, we'll need to modify this and the other places the values are updated to something like `self.owcurs = self.owcurs.wrapping_add(2)`.  If special remediation is required at the point of wrap-around, then https://doc.rust-lang.org/std/primitive.usize.html#method.overflowing_add can be used to detect when that happens.

Reply:
This is dealt with in the `race-buffer-modifications` branch